agent/xds/routes.go

```diff
@@ -430,6 +425,75 @@ func (s *ResourceGenerator) routesForIngressGateway(cfgSnap *proxycfg.ConfigSnap
 	return result, nil
 }
 
+// routesForAPIGateway returns the xDS API representation of the
+// "routes" in the snapshot.
+func (s *ResourceGenerator) routesForAPIGateway(cfgSnap *proxycfg.ConfigSnapshot) ([]proto.Message, error) {
+	var result []proto.Message
+
+	readyUpstreamsList := getReadyUpstreams(cfgSnap)
+
+	for _, readyUpstreams := range readyUpstreamsList {
+		listenerCfg := readyUpstreams.listenerCfg
+		// Do not create any route configuration for TCP listeners
+		if listenerCfg.Protocol == "tcp" {
+			continue
+		}
+
+		routeRef := readyUpstreams.routeReference
+		listenerKey := readyUpstreams.listenerKey
+
+		// Depending on their TLS config, upstreams are either attached to the
+		// default route or have their own routes. We'll add any upstreams that
+		// don't have custom filter chains and routes to this.
+		defaultRoute := &envoy_route_v3.RouteConfiguration{
+			Name: listenerKey.RouteName(),
+			// ValidateClusters defaults to true when defined statically and false
+			// when done via RDS. Re-set the reasonable value of true to prevent
+			// null-routing traffic.
+			ValidateClusters: makeBoolValue(true),
+		}
+
+		route, ok := cfgSnap.APIGateway.HTTPRoutes.Get(routeRef)
```

Comment:
are the HTTPRoutes here flattened previously or are they the routes that we're subscribing to?

Reply:
These are the routes we are subscribing here, they do not seem to be flattened when we get them here.

Reply:
Correct, this is just the collection of HTTPRoutes that we [watch](https://github.com/hashicorp/consul/blob/e00280e7df8034c8c8bd48498e76a98754d63a1c/agent/proxycfg/api_gateway.go#L173) and then [set](https://github.com/hashicorp/consul/blob/e00280e7df8034c8c8bd48498e76a98754d63a1c/agent/proxycfg/api_gateway.go#L292) while crawling the dependency graph